cmd/notation/plugin/install.go

```diff
@@ -0,0 +1,329 @@
+)
+
+type pluginInstallOpts struct {
+	cmd.LoggingFlagOpts
+	pluginSourceType notationplugin.PluginSourceType
+	pluginSource     string
+	inputCheckSum    string
+	isFile           bool
+	isUrl            bool
+	force            bool
+}
+
+func pluginInstallCommand(opts *pluginInstallOpts) *cobra.Command {
+	if opts == nil {
+		opts = &pluginInstallOpts{}
+	}
+	command := &cobra.Command{
+		Use:     "install [flags] <--file|--url> <plugin_source>",
+		Aliases: []string{"add"},
+		Short:   "Install plugin",
+		Long: `Install a plugin
+
+Example - Install plugin from file system:
+  notation plugin install --file wabbit-plugin-v1.0.zip
+
+Example - Install plugin from file system with user input SHA256 checksum:
+  notation plugin install --file wabbit-plugin-v1.0.zip --checksum abcdef 
+
+Example - Install plugin from file system regardless if it's already installed:
+  notation plugin install --file wabbit-plugin-v1.0.zip --force
+
+Example - Install plugin from file system with .tar.gz:
+  notation plugin install --file wabbit-plugin-v1.0.tar.gz
+
+Example - Install plugin from URL, SHA256 checksum is required:
+  notation plugin install --url https://wabbit-networks.com/intaller/linux/amd64/wabbit-plugin-v1.0.tar.gz --checksum abcxyz
+`,
+		Args: func(cmd *cobra.Command, args []string) error {
+			if len(args) == 0 {
+				return errors.New("missing plugin file path or URL")
+			}
+			opts.pluginSource = args[0]
+			return nil
+		},
+		PreRunE: func(cmd *cobra.Command, args []string) error {
+			if opts.isFile {
+				opts.pluginSourceType = notationplugin.PluginSourceTypeFile
+				return nil
+			}
+			if opts.isUrl {
+				opts.pluginSourceType = notationplugin.PluginSourceTypeURL
+				return nil
+			}
+			return errors.New("must choose one and only one flag from [--file, --url]")
+		},
+		RunE: func(cmd *cobra.Command, args []string) error {
+			return installPlugin(cmd, opts)
+		},
+	}
+	opts.LoggingFlagOpts.ApplyFlags(command.Flags())
+	command.Flags().BoolVar(&opts.isFile, "file", false, "install plugin from a file in file system")
+	command.Flags().BoolVar(&opts.isUrl, "url", false, "install plugin from an HTTPS URL")
+	command.Flags().StringVar(&opts.inputCheckSum, "sha256sum", "", "must match SHA256 of the plugin source")
+	command.Flags().BoolVar(&opts.force, "force", false, "force the installation of a plugin")
+	command.MarkFlagsMutuallyExclusive("file", "url")
+	return command
+}
+
+func installPlugin(command *cobra.Command, opts *pluginInstallOpts) error {
+	// set log level
+	ctx := opts.LoggingFlagOpts.InitializeLogger(command.Context())
+	// core process
+	switch opts.pluginSourceType {
+	case notationplugin.PluginSourceTypeFile:
+		return installFromFileSystem(ctx, opts.pluginSource, opts.inputCheckSum, opts.force)
+	case notationplugin.PluginSourceTypeURL:
+		if opts.inputCheckSum == "" {
+			return errors.New("install from URL requires non-empty SHA256 checksum of the plugin source")
```

Comment:
There's one thing about `PreRunE`, it is evaluated before the check of `ValidateFlagGroups`, i.e. before the flags check set by `MarkFlagsMutuallyExclusive` and `MarkFlagsOneRequired`. Since `checksum` is required by `--url` and optional by `--file`, checking `checksum` in `PreRunE` complicates the logic comparing to checking it after the validation of the flags.